File: internal/stream/stream.go
// Package stream is m6t's loopback WebSocket transport (DESIGN.md §3.3): the
// channel that carries throughput-sensitive data between the Go backend and the
// webview, starting with PTY I/O.
//
// The Wails bridge handles RPC — open a file, run an apply, list projects. It is
// the wrong shape for a terminal, where a build log arrives as thousands of
// small writes and every one of them would cross a JSON marshaling boundary.
// So the backend also serves an HTTP listener on 127.0.0.1 with a random port
// and a per-launch bearer token, and the frontend opens sockets to it. The only
// thing the bridge carries is the Endpoint that says where and with what token.
//
// The wire protocol — frame kinds, the control envelope, the auth forms and the
// backpressure rule — is specified in PROTOCOL.md next to this file. It is a
// contract with the frontend, so it is written down rather than inferred from
// the handlers.
//
// This package spawns nothing and owns no process. It takes a Terminals seam in
// its constructor and moves bytes across it: the PTY service on the other side
// is a sibling that this package must not import, and the binding layer is what
// joins the two (CLAUDE.md, "Architecture map").
package stream

import (
	"errors"
	"time"
)

const (
	// loopbackAddr binds the listener. The interface is not configurable on
	// purpose: a stream server reachable from another host is a shell server,
	// and port 0 means the OS picks a free port so two m6t instances can run
	// side by side.
	loopbackAddr = "127.0.0.1:0"

	// protocolVersion is the WebSocket subprotocol the server negotiates. A
	// browser that offers subprotocols requires the server to select one, so
	// this is what gets echoed back when the token arrives as a subprotocol.
	protocolVersion = "m6t.v1"

	// authSubprotocolPrefix carries the bearer credential for clients that
	// cannot set request headers. The browser WebSocket API is one: it exposes
	// the subprotocol list and nothing else, which is the same reason the
	// Kubernetes API server accepts a credential this way.
	//
	// The identifier deliberately avoids the word this prefix contains: gosec
	// G101 matches on names, and a constant named for a credential — even one
	// holding a fixed, public prefix rather than a secret — is a finding this
	// repo has no way to record except by suppressing it.
	authSubprotocolPrefix = "m6t.token."

	// bearerPrefix is the Authorization header form, used by every client that
	// can set headers — the Go tests, and any tooling attached to a session.
	bearerPrefix = "Bearer "

	// wailsScheme and wailsHost are the origins the Wails webview reports:
	// wails://wails on macOS and Linux, http://wails.localhost on Windows.
	wailsScheme = "wails"
	wailsHost   = "wails.localhost"

	// localhostHost is the loopback name that does not parse as an IP.
	localhostHost = "localhost"

	// outboundQueue bounds the frames one connection may have waiting. It is
	// the whole of the backpressure policy: a client this far behind starts
	// losing frames rather than being allowed to slow the producer down. At
	// PTY chunk size that is a couple of megabytes of slack, which absorbs a
	// render stall without absorbing a wedged webview.
	outboundQueue = 64

	// maxMessageBytes bounds one inbound message. Terminal input is keystrokes
	// and pastes and control frames are tiny, so anything approaching this is
	// either a bug or an attempt to make the backend allocate.
	maxMessageBytes = 1 << 20

	// readHeaderTimeout bounds how long a connection may take to send its
	// request headers. A local client is immediate; the timeout is what stops
	// an opened-and-abandoned socket from holding a handler forever.
	readHeaderTimeout = 5 * time.Second

	// shutdownTimeout bounds the graceful close on the way out. The app is
	// quitting, so a connection that will not finish is dropped rather than
	// allowed to delay it.
	shutdownTimeout = 2 * time.Second

	// sessionIDParam is the path wildcard naming the terminal session.
	sessionIDParam = "sessionID"
)

// errNotStarted reports an Endpoint asked for before the listener is up.
var errNotStarted = errors.New("stream server is not started")

// Endpoint is what a frontend needs to open a socket: the port the listener
// actually got, and the token every connection must present.
//
// It crosses the Wails bridge, and it is the one piece of this package that
// must never be written to a log or an error message — a token in a log file
// outlives the launch it was minted for.
type Endpoint struct {
	Port  int    `json:"port"`
	Token string `json:"token"`
}

// Attachment is one consumer's view of a terminal session, in the terms this
// package needs: the scrollback to replay, the output that follows, how the
// child ended, and how to stop consuming.
//
// It mirrors the PTY service's own attachment rather than reusing it. Sibling
// services do not import each other, so the shape is restated here and the
// binding layer adapts one to the other — the cost of the seam, paid once.
type Attachment struct {
	// Replay is the scrollback at the moment of attaching.
	Replay []byte

	// Chunks carries output produced after the attach, and closes when the
	// session ends or the consumer detaches.
	Chunks <-chan []byte

	// Exited yields the child's exit code once. It closes without yielding
	// when the consumer detached before the child exited.
	Exited <-chan int

	// Detach releases the consumer. It must not be nil and it must be
	// idempotent: the server calls it on every path out of a connection,
	// including the one where the session had already ended.
	Detach func()
}

// Terminals is the PTY service as this package uses it. The methods take an
// opaque session identifier because that is all a transport needs to know: the
// server routes bytes to a name the frontend was given, and what a session is
// stays behind this seam.
type Terminals interface {
	Attach(id string) (Attachment, error)
	Write(id string, p []byte) error
	Resize(id string, cols, rows uint16) error
	Kill(id string) error
}

File: frontend/src/lib/stream.ts
import type { stream } from "../../wailsjs/go/models";

/**
 * Discovery and connection for the loopback stream server
 * (`internal/stream/PROTOCOL.md` §1–§2).
 *
 * The port and token are minted per launch and reach the frontend over the
 * Wails bridge and nowhere else, so nothing here logs them — an endpoint in a
 * console line outlives the launch it was minted for.
 */

/**
 * Where the stream server is listening, and the credential it requires.
 *
 * Aliased from the generated binding rather than restated, so a change to the
 * Go struct fails type-checking here instead of silently disagreeing.
 */
export type Endpoint = stream.Endpoint;

/**
 * The subprotocol the server selects. A browser that offers subprotocols
 * requires the server to select one, so the version travels alongside the
 * credential rather than being implied (§2).
 */
const VERSION_SUBPROTOCOL = "m6t.v1";

/**
 * Prefix that carries the bearer credential. The browser WebSocket API cannot
 * set request headers, so the token rides in the subprotocol list — the same
 * mechanism the Kubernetes API server accepts a credential through.
 */
const TOKEN_SUBPROTOCOL_PREFIX = "m6t.token.";

/** The socket URL for one terminal session (§4). */
export function terminalURL(endpoint: Endpoint, sessionID: string): string {
  return `ws://127.0.0.1:${String(endpoint.port)}/pty/${encodeURIComponent(sessionID)}`;
}

/** The subprotocol list every connection must offer (§2). */
export function subprotocols(endpoint: Endpoint): string[] {
  return [VERSION_SUBPROTOCOL, TOKEN_SUBPROTOCOL_PREFIX + endpoint.token];
}

/** How a socket is constructed, injectable so tests need no live server. */
export type SocketFactory = (url: string, protocols: string[]) => WebSocket;

const defaultSocketFactory: SocketFactory = (url, protocols) =>
  new WebSocket(url, protocols);

/** Opens a WebSocket to a terminal session, ready to carry binary frames. */
export function openTerminalSocket(
  endpoint: Endpoint,
  sessionID: string,
  create: SocketFactory = defaultSocketFactory,
): WebSocket {
  const socket = create(
    terminalURL(endpoint, sessionID),
    subprotocols(endpoint),
  );
  // PTY output is raw bytes. Without this the runtime delivers them as Blobs,
  // which are only readable asynchronously — long enough to let a later control
  // frame overtake the output it was meant to follow.
  socket.binaryType = "arraybuffer";
  return socket;
}
